From 8afd2a389ba10c78722f6db99b3b169e082e4125 Mon Sep 17 00:00:00 2001
From: Hansong Zhang <hsz@google.com>
Date: Mon, 14 Jan 2019 14:59:35 -0800
Subject: [PATCH 1/2] process_l2cap_cmd: Fix OOB

Bug: 119870451
Test: POC
Change-Id: I2f5e7fedd9aed96c4ffc55af79fdac61c2e5b087
Merged-In: I5131bbf9cda6248fdbbc4bb91916b2fe3731246e
(cherry picked from commit 94fd011bc9a72081cc691ed7d6e6eec42e9f4539)
---
 stack/l2cap/l2c_main.c  | 72 ++++++++++++++++++++++++++++++++++++++++---------
 stack/l2cap/l2c_utils.c |  3 +++
 2 files changed, 63 insertions(+), 12 deletions(-)

diff --git a/stack/l2cap/l2c_main.c b/stack/l2cap/l2c_main.c
index 5eced53..6d5c390 100644
--- a/stack/l2cap/l2c_main.c
+++ b/stack/l2cap/l2c_main.c
@@ -490,7 +490,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
                 {
                 case L2CAP_CFG_TYPE_MTU:
                     cfg_info.mtu_present = TRUE;
-                    if (p + 2 > p_next_cmd) {
+                    if (cfg_len != 2) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -499,7 +503,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_FLUSH_TOUT:
                     cfg_info.flush_to_present = TRUE;
-                    if (p + 2 > p_next_cmd) {
+                    if (cfg_len != 2) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -508,7 +516,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_QOS:
                     cfg_info.qos_present = TRUE;
-                    if (p + 2 + 5 * 4 > p_next_cmd) {
+                    if (cfg_len != 2 + 5 * 4) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -523,7 +535,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_FCR:
                     cfg_info.fcr_present = TRUE;
-                    if (p + 3 + 3 * 2 > p_next_cmd) {
+                    if (cfg_len != 3 + 3 * 2) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -537,7 +553,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_FCS:
                     cfg_info.fcs_present = TRUE;
-                    if (p + 1 > p_next_cmd) {
+                    if (cfg_len != 1) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -546,7 +566,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_EXT_FLOW:
                     cfg_info.ext_flow_spec_present = TRUE;
-                    if (p + 2 + 2 + 3 * 4 > p_next_cmd) {
+                    if (cfg_len != 2 + 2 + 3 * 4) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -629,7 +653,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
                 {
                 case L2CAP_CFG_TYPE_MTU:
                     cfg_info.mtu_present = TRUE;
-                    if (p + 2 > p_next_cmd) {
+                    if (cfg_len != 2) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -638,7 +666,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_FLUSH_TOUT:
                     cfg_info.flush_to_present = TRUE;
-                    if (p + 2 > p_next_cmd) {
+                    if (cfg_len != 2) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -647,7 +679,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_QOS:
                     cfg_info.qos_present = TRUE;
-                    if (p + 2 + 5 * 4 > p_next_cmd) {
+                    if (cfg_len != 2 + 5 * 4) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -662,7 +698,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_FCR:
                     cfg_info.fcr_present = TRUE;
-                    if (p + 3 + 3 * 2 > p_next_cmd) {
+                    if (cfg_len != 3 + 3 * 2) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -676,7 +716,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_FCS:
                     cfg_info.fcs_present = TRUE;
-                    if (p + 1 > p_next_cmd) {
+                    if (cfg_len != 1) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
@@ -685,7 +729,11 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
 
                 case L2CAP_CFG_TYPE_EXT_FLOW:
                     cfg_info.ext_flow_spec_present = TRUE;
-                    if (p + 2 + 2 + 3 * 4 > p_next_cmd) {
+                    if (cfg_len != 2 + 2 + 3 * 4) {
+                      android_errorWriteLog(0x534e4554, "119870451");
+                      return;
+                    }
+                    if (p + cfg_len > p_next_cmd) {
                       android_errorWriteLog(0x534e4554, "74202041");
                       return;
                     }
diff --git a/stack/l2cap/l2c_utils.c b/stack/l2cap/l2c_utils.c
index 58ac4b0..72bd52e 100644
--- a/stack/l2cap/l2c_utils.c
+++ b/stack/l2cap/l2c_utils.c
@@ -859,6 +859,9 @@ void l2cu_send_peer_config_rej (tL2C_CCB *p_ccb, UINT8 *p_data, UINT16 data_len,
             case L2CAP_CFG_TYPE_MTU:
             case L2CAP_CFG_TYPE_FLUSH_TOUT:
             case L2CAP_CFG_TYPE_QOS:
+            case L2CAP_CFG_TYPE_FCR:
+            case L2CAP_CFG_TYPE_FCS:
+            case L2CAP_CFG_TYPE_EXT_FLOW:
                 p_data += cfg_len + L2CAP_CFG_OPTION_OVERHEAD;
                 break;
 
-- 
2.7.4

